File: CPP_Programmes/infixToPostfix.cpp
#include<iostream>
#include<stdio.h>
#include<ctype.h>
#include<string.h>
using namespace std;
#define MAX 20000
char infix[MAX];
char postfix[MAX];
char stack[MAX];
char choice;
int top=-1;
int i=0;int j=0;
int precedence(char choice)
{
    switch (choice)
    {
        case '^':return 3; break;
        case '*':case '/': return 2; break;
        case '+':case '-':return 1; break;
        default: return 0; break;
    }
}
void push(char item)
{
        stack[++top]=item;
}
char pop()
{
        return stack[top--];
}
void InfixToPostfix()
{
while(infix[i]!='\0')
{
    choice=infix[i];
    if(isalnum(choice))
    {
        postfix[j]=choice;
        postfix[++j]='\0';
    }
    else if(choice=='(')
    {
        push(choice);
    }
    else if(choice==')')
    {
        char a;
        while((a=pop())!='(')
        {
        postfix[j]=a;
        postfix[++j]='\0';
        }
    }
    else
    {
        while (precedence(stack[top])>=precedence(choice))
        {
            postfix[j]=pop();
        postfix[++j]='\0';
            }push(choice);
    }
    i++;

}
cout<<"The postfix expression is "<<postfix<<endl;
}
int main()
{
    cout<<"Enter an infix expression"<<endl;
    cin>>infix;
    int len=strlen(infix);
    infix[len]=')';
    infix[++len]='\0';
    push('(');
    InfixToPostfix();
    return 0;
}

File: Python3_Programmes/infixToPostfix.py
stack = []
postfix = []
infix=""
i=0


def precedence(choice):
    if(choice=='^'):
        return 3
    elif(choice=='*'or choice=='/'):
        return 2
    elif(choice=='+' or choice=='-'):
        return 2
    else:
        return 0
    

def InfixToPostfix():
    global infix
    global choice
    global i
    while(infix[i]!='d'):
        choice=infix[i];
        if(choice==' '):
            i+=1
            continue
        elif(choice.isalnum()):
            postfix.append(choice)
        elif(choice=='('):
            stack.append(choice)
        elif(choice==')'):
            
            while(stack[-1]!='('):
                postfix.append(stack.pop())
            stack.pop()
        else:
            while(precedence(stack[-1])>=precedence(choice)):
                postfix.append(stack.pop())
            stack.append(choice)
        i+=1
    res=''
    for i in postfix:
        res+=i
    print(f"The postfix expression is {res[0:-2]}")
    return
    
    
def main():
    global infix
    infix = input("Enter an infix expression\n")
    len_str=len(infix)
    print(f"{infix}")
    infix = infix+')'
    infix = infix+'end'
    stack.append('(')
    InfixToPostfix()

if __name__=='__main__':
    main()
